File: js/src/acp.ts
/**
 * ACP (Agent Client Protocol) model source — issue #96, ADR 0031
 * ("ACP: the warm session is the feature, and it breaks the stateless request"),
 * openspec/changes/add-acp-model-source.
 *
 * ACP is to *agents* what MCP is to *tools*: JSON-RPC 2.0, one object per line,
 * over a child process's stdin/stdout (`devin acp`, Gemini CLI, Zed's agents all
 * speak it). This file ships ACP as nothing more than a `generate` function —
 * the exact shape {@link InProcessOptions.generate} already defines — so the
 * tool-calling loop, skills, MCP tools, adapters and sub-agents are entirely
 * untouched. `loadACP` opens the connection once (`initialize` → `session/new` →
 * optional `session/set_mode`) and hands back an {@link ACPClient} whose
 * `.generate` can be passed straight to {@link createInProcessClient}.
 *
 * The hard part is not the JSON-RPC plumbing — it is that an ACP session is
 * *stateful* while every other toolnexus model source is stateless-by-default.
 * toolnexus assembles a complete request every turn; sending the whole thing
 * into a session that already remembers the transcript makes some agents answer
 * a near-duplicate prompt from stale history (reproduced in `spikes/acp/`).
 * The mitigation (kept as the default, not left to the caller) is an explicit
 * "this supersedes everything earlier" marker appended to every rendered
 * prompt — see {@link renderPrompt}.
 *
 * Three more traps this client exists to close, all proven in the spike:
 *  - `session/update` notifications interleave with JSON-RPC replies on the
 *    same stdout stream, so everything arriving is demultiplexed by id.
 *  - Only `agent_message_chunk` text may be accumulated into the reply —
 *    `agent_thought_chunk` and tool-call narration wrap prose around
 *    structured output and corrupt it if mixed in.
 *  - An unanswered `session/request_permission` hangs the turn FOREVER, even
 *    in an agent's bypass mode. The client answers with the first `allow`-kind
 *    option the instant the request arrives; `permissionTimeoutMs` is only a
 *    safety net for a broken agent that never asks and never replies at all.
 */
import { spawn } from "node:child_process"
import { createInterface } from "node:readline"
import path from "node:path"
import type { InProcessRequest, InProcessResponse } from "./client.js"

/** Options for {@link loadACP}. `command`/`args` spawn the ACP agent as a child
 *  process; everything else tunes the session handshake. */
export interface ACPOptions {
  /** The ACP agent binary, e.g. `"devin"`. */
  command: string
  /** Arguments to the agent, e.g. `["acp"]`. */
  args?: string[]
  /** Working directory for `session/new`. Real ACP agents (e.g. `devin acp`)
   *  reject a relative `cwd` with `-32602 Invalid params` — this is always
   *  resolved to an absolute path before it is sent. Default: `process.cwd()`. */
  cwd?: string
  /** Extra environment variables for the child, merged over the current
   *  process environment. */
  env?: Record<string, string>
  /** ACP `initialize` protocolVersion. Default `1`. */
  protocolVersion?: number
  /** Safety-net timeout (ms) for one `session/prompt` turn — NOT specifically
   *  for the permission handshake, which this client always answers instantly.
   *  It exists only so a broken/unresponsive agent cannot hang a caller
   *  forever. Default 30000. */
  permissionTimeoutMs?: number
  /** Optional `session/set_mode` sent once, right after `session/new`. */
  mode?: string
}

/** One warm ACP connection: one child process, one session, alive across many
 *  turns. `.generate` has exactly the shape {@link InProcessOptions.generate}
 *  wants, so it plugs straight into {@link createInProcessClient}. */
export interface ACPClient {
  generate: (request: InProcessRequest) => Promise<InProcessResponse>
  /** Terminates the agent process. Idempotent — a second call is a no-op. */
  close: () => Promise<void>
}

/** The literal marker every rendered prompt ends with, naming the current
 *  turn as the one to answer. Kept as a named constant (not just inlined)
 *  because the exact wording is shared, by convention, with the other
 *  language ports and the spike that proved it fixes the stale-answer bug. */
const SUPERSEDES_MARKER = "SUPERSEDES-ALL-PRIOR:"

interface RpcMessage {
  jsonrpc?: string
  id?: string | number
  method?: string
  params?: any
  result?: any
  error?: { code: number; message: string }
}

function renderContentPart(part: any): string {
  if (typeof part === "string") return part
  if (part == null) return ""
  if (typeof part === "object" && "text" in part && typeof part.text === "string") return part.text
  return JSON.stringify(part)
}

/** Flattens `content` (a string, or an array of content parts) to plain text. */
function renderContent(content: any): string {
  if (typeof content === "string") return content
  if (Array.isArray(content)) return content.map(renderContentPart).join(" ")
  return renderContentPart(content)
}

/** Renders the FULL assembled request as one prompt string — role + content
 *  per message — then appends the supersedes marker naming the latest user
 *  turn, per ADR 0031's proposed (and spike-confirmed) default: full request
 *  every turn, not a delta, so this stays stateless like every other
 *  toolnexus model source and never risks a shadow transcript drifting from
 *  the caller's own conversation state. */
function renderPrompt(request: InProcessRequest): string {
  const messages = request.messages ?? []
  const lines = messages.map((m) => `${m?.role ?? "user"}: ${renderContent(m?.content)}`)

  let latest = ""
  for (let i = messages.length - 1; i >= 0; i--) {
    if (messages[i]?.role === "user") {
      latest = renderContent(messages[i].content)
      break
    }
  }
  if (!latest && messages.length > 0) latest = renderContent(messages[messages.length - 1]?.content)

  lines.push(`${SUPERSEDES_MARKER} ${latest}`)
  return lines.join("\n")
}

/**
 * Opens one ACP connection: spawns `command`, performs `initialize` →
 * `session/new` (with an absolute `cwd` and an `mcpServers` array — the exact
 * trap real `devin acp` enforces with `-32602` otherwise) → an optional
 * `session/set_mode`, then returns a client whose `.generate` sends exactly
 * one `session/prompt` per call, accumulating only `agent_message_chunk`
 * text.
 */
export async function loadACP(opts: ACPOptions): Promise<ACPClient> {
  const cwd = opts.cwd ? path.resolve(opts.cwd) : process.cwd()
  const permissionTimeoutMs = opts.permissionTimeoutMs ?? 30000

  const child = spawn(opts.command, opts.args ?? [], {
    cwd,
    env: opts.env ? { ...process.env, ...opts.env } : process.env,
    stdio: ["pipe", "pipe", "ignore"],
  })

  let closed = false
  let nextId = 0
  const pending = new Map<string, (msg: RpcMessage) => void>()
  let updateSink: ((params: any) => void) | null = null

  const rl = createInterface({ input: child.stdout! })
  rl.on("line", (raw) => {
    const line = raw.trim()
    if (!line) return
    let msg: RpcMessage
    try {
      msg = JSON.parse(line)
    } catch {
      return
    }
    handleLine(msg)
  })

  function write(msg: Record<string, unknown>): void {
    child.stdin!.write(JSON.stringify(msg) + "\n")
  }

  function reply(id: unknown, result: unknown): void {
    write({ jsonrpc: "2.0", id, result })
  }

  function call(method: string, params: unknown): Promise<any> {
    const id = `c-${++nextId}`
    return new Promise((resolve, reject) => {
      pending.set(id, (msg) => {
        if (msg.error) reject(new Error(`toolnexus: ACP error ${msg.error.code}: ${msg.error.message}`))
        else resolve(msg.result)
      })
      write({ jsonrpc: "2.0", id, method, params })
    })
  }

  /** Demultiplexes EVERYTHING arriving on stdout: replies to our requests
   *  (matched by JSON-RPC id), `session/update` notifications (interleaved
   *  with replies, routed to whichever turn is in flight), and
   *  server-initiated requests like `session/request_permission`. */
  function handleLine(msg: RpcMessage): void {
    if (msg.method === undefined && msg.id !== undefined) {
      const cb = pending.get(String(msg.id))
      if (cb) {
        pending.delete(String(msg.id))
        cb(msg)
      }
      return
    }
    if (msg.method === "session/update") {
      updateSink?.(msg.params)
      return
    }
    if (msg.method === "session/request_permission" && msg.id !== undefined) {
      answerPermission(msg)
      return
    }
    // An unhandled server->client request still needs a reply or a
    // strict agent may itself hang waiting on one.
    if (msg.id !== undefined) reply(msg.id, {})
  }

  /** Answers with the FIRST option whose `kind` starts with `allow` — an
   *  unanswered permission request hangs the turn forever (proven in
   *  spikes/acp). This runs the instant the request arrives, never waiting
   *  for anything else. */
  function answerPermission(req: RpcMessage): void {
    const options: Array<{ optionId: string; kind?: string }> = req.params?.options ?? []
    const chosen = options.find((o) => typeof o.kind === "string" && o.kind.startsWith("allow"))
    const result = chosen
      ? { outcome: { outcome: "selected", optionId: chosen.optionId } }
      : { outcome: { outcome: "cancelled" } }
    reply(req.id, result)
  }

  await call("initialize", {
    protocolVersion: opts.protocolVersion ?? 1,
    clientCapabilities: { fs: { readTextFile: false, writeTextFile: false } },
  })

  const sessionResult = await call("session/new", { cwd, mcpServers: [] })
  const sessionId = sessionResult?.sessionId

  if (opts.mode) {
    await call("session/set_mode", { sessionId, modeId: opts.mode })
  }

  // Serialises turns: one ACP session is one conversation, so concurrent
  // prompts must not interleave into it. Idiomatic promise-chaining queue —
  // each generate() call attaches after whatever is already queued, and the
  // queue itself never rejects (a failed turn does not poison later ones).
  let turnQueue: Promise<unknown> = Promise.resolve()

  async function promptOnce(request: InProcessRequest): Promise<InProcessResponse> {
    const text = renderPrompt(request)
    const id = `c-${++nextId}`

    let content = ""
    updateSink = (params: any) => {
      const update = params?.update
      if (update?.sessionUpdate === "agent_message_chunk") {
        content += update?.content?.text ?? ""
      }
      // agent_thought_chunk and tool_call/tool_call_update narration are
      // deliberately dropped — mixing them into the reply corrupts
      // structured output a caller expects to parse.
    }

    const replyPromise = new Promise<RpcMessage>((resolve) => pending.set(id, resolve))
    write({ jsonrpc: "2.0", id, method: "session/prompt", params: { sessionId, prompt: [{ type: "text", text }] } })

    let timer: NodeJS.Timeout | undefined
    const timeout = new Promise<never>((_, reject) => {
      timer = setTimeout(
        () => reject(new Error(`toolnexus: ACP turn timed out after ${permissionTimeoutMs}ms`)),
        permissionTimeoutMs,
      )
    })

    try {
      const msg = await Promise.race([replyPromise, timeout])
      if (msg.error) throw new Error(`toolnexus: ACP error ${msg.error.code}: ${msg.error.message}`)
      return { content }
    } finally {
      clearTimeout(timer)
      pending.delete(id)
      updateSink = null
    }
  }

  const generate = (request: InProcessRequest): Promise<InProcessResponse> => {
    if (closed) return Promise.reject(new Error("toolnexus: ACP client is closed"))
    const turn = turnQueue.then(() => promptOnce(request))
    turnQueue = turn.then(
      () => undefined,
      () => undefined,
    )
    return turn
  }

  const close = (): Promise<void> => {
    if (closed) return Promise.resolve()
    closed = true
    rl.close()
    return new Promise((resolve) => {
      if (child.exitCode !== null || child.signalCode !== null) {
        resolve()
        return
      }
      const done = () => resolve()
      child.once("exit", done)
      child.kill()
      // Belt-and-braces: resolve even if the child never emits `exit` (e.g.
      // already reaped), so close() itself never hangs.
      setTimeout(done, 1000)
    })
  }

  const client: ACPClient = { generate, close }
  // Non-enumerable, undocumented — internal test hook only, so the public
  // ACPClient surface stays exactly {generate, close}.
  Object.defineProperty(client, "_pid", { value: child.pid, enumerable: false })
  return client
}

File: golang/acp.go
// ACP (Agent Client Protocol) model source (ADR 0025, issue #96).
//
// ACP is to *agents* what MCP is to *tools*: JSON-RPC 2.0, one object per
// line, over a child process's stdin/stdout — the exact framing local stdio
// MCP already uses in mcp.go. This file ships ACP as a `Generate` source
// (LoadACP(...).Generate has the exact shape of InProcessOptions.Generate,
// see inprocess.go), so the tool-calling loop, skills, MCP, adapters and
// sub-agents are untouched. It is NOT a new tool source and NOT a new
// client — see ADR 0025 and openspec/changes/add-acp-model-source.
//
// The warm session is the feature (ADR 0025's measurements): the agent
// process and its ACP session are created once by LoadACP and then serve
// every turn as a `session/prompt` on that same session — amortising the
// process-startup cost that dominates a cold `devin -p` call.
//
// The hard part is that toolnexus assembles a COMPLETE request every turn
// (the full message array), while an ACP session is STATEFUL — it already
// has the transcript. Sending the whole thing again each turn makes a naive
// agent answer a stale, near-duplicate prompt from its own history. The
// mitigation (default, per ADR 0025): render the full transcript into the
// prompt text every turn, and append an explicit "this supersedes every
// earlier prompt" marker naming the latest turn — see renderACPPrompt.
//
// Everything else is mechanical: demultiplex stdout by JSON-RPC id because
// `session/update` notifications interleave with our own request replies;
// accumulate ONLY `agent_message_chunk` text (thoughts and tool narration
// must be dropped or they corrupt structured output the host expects back);
// answer `session/request_permission` with the first `allow`-kind option —
// an unanswered permission request hangs the turn forever, even in bypass
// mode; serialise turns on one session (one ACP session is one conversation,
// concurrent prompts must not interleave into one transcript); and keep the
// child process's lifetime independent of any one turn's cancellation.
package toolnexus

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"sync"
	"time"
)

// acpSupersedesMarker names the current/latest user turn as superseding all
// earlier prompts in the (stateful) ACP session's own transcript. The exact
// text is a client-side convention, not part of the ACP protocol; kept
// recognizable and stable so an agent (or a test fixture) can key off it.
const acpSupersedesMarker = "SUPERSEDES-ALL-PRIOR:"

// ---------------------------------------------------------------------------
// Wire types (JSON-RPC 2.0, one object per line — identical framing to local
// stdio MCP in mcp.go, just without an SDK: ACP has no mature Go client yet).
// ---------------------------------------------------------------------------

type acpMsg struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      json.RawMessage `json:"id,omitempty"`
	Method  string          `json:"method,omitempty"`
	Params  json.RawMessage `json:"params,omitempty"`
	Result  json.RawMessage `json:"result,omitempty"`
	Error   *acpError       `json:"error,omitempty"`
}

type acpError struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
}

// ACPPermissionTimeoutError is returned when a turn is still blocked on an
// unanswered `session/request_permission` after ACPOptions.PermissionTimeout
// — the safety net for a broken/misbehaving agent. The normal path never
// hits this: a well-formed permission request is auto-answered the moment it
// arrives (see ACPClient.readLoop), so the turn completes immediately after.
type ACPPermissionTimeoutError struct{ Waited time.Duration }

func (e *ACPPermissionTimeoutError) Error() string {
	return fmt.Sprintf("toolnexus: acp: turn hung waiting on session/request_permission for %s", e.Waited)
}

// ---------------------------------------------------------------------------
// ACPOptions / LoadACP
// ---------------------------------------------------------------------------

// ACPOptions configures the child ACP agent process and the session opened
// against it. Only Command is required.
type ACPOptions struct {
	// Command is the ACP agent binary to spawn (e.g. "devin", with
	// Args: []string{"acp"}).
	Command string
	Args    []string
	// Env is appended to the child's environment (which otherwise inherits
	// this process's environment, exactly like exec.Command's default).
	Env []string

	// Cwd is the absolute working directory handed to `session/new`. A real
	// `devin acp` REJECTS session/new with -32602 without an absolute cwd
	// (ADR 0025) — if left empty, LoadACP resolves os.Getwd() and requires
	// that it is absolute (it always is on every supported OS).
	Cwd string

	// ProtocolVersion is negotiated in `initialize`. Defaults to 1.
	ProtocolVersion int

	// Mode, if set, issues an optional `session/set_mode` after
	// `session/new` (e.g. an agent's "bypass permissions" or "plan" mode).
	Mode string

	// PermissionTimeout bounds how long a turn waits on an unanswered
	// session/request_permission before returning ACPPermissionTimeoutError.
	// Defaults to 30s. This is a safety net for a broken agent, not the
	// normal path: a well-formed request is answered immediately.
	PermissionTimeout time.Duration

	// RequestTimeout bounds `initialize` / `session/new` / `session/set_mode`
	// and the reply half of `session/prompt`. Defaults to 30s.
	RequestTimeout time.Duration
}

// ACPClient is one warm ACP session: one child process, one session id,
// alive across many turns. Build one with LoadACP; its Generate method has
// the exact shape InProcessOptions.Generate expects.
type ACPClient struct {
	cmd    *exec.Cmd
	stdin  io.WriteCloser
	stdout *bufio.Scanner

	mu      sync.Mutex // guards pending + nextID + updateSink
	nextID  int64
	pending map[string]chan acpMsg

	sessionID string

	// promptMu serialises turns: one ACP session is one conversation, and
	// concurrent session/prompt calls would interleave into one transcript.
	promptMu sync.Mutex

	updateSink func(json.RawMessage)

	permissionTimeout time.Duration
	requestTimeout    time.Duration

	closeOnce sync.Once
	closeErr  error
}

// LoadACP spawns the ACP agent as a child process, completes `initialize` +
// `session/new` (+ optional `session/set_mode`), and returns a live,
// warm ACPClient. ctx bounds only this setup handshake — once LoadACP
// returns, the child process's lifetime is independent of ctx (ADR 0025:
// "process lifetime independent of any one turn's cancellation").
func LoadACP(ctx context.Context, opts ACPOptions) (*ACPClient, error) {
	if opts.Command == "" {
		return nil, fmt.Errorf("toolnexus: acp: ACPOptions.Command is required")
	}

	protocolVersion := opts.ProtocolVersion
	if protocolVersion == 0 {
		protocolVersion = 1
	}
	permissionTimeout := opts.PermissionTimeout
	if permissionTimeout <= 0 {
		permissionTimeout = 30 * time.Second
	}
	requestTimeout := opts.RequestTimeout
	if requestTimeout <= 0 {
		requestTimeout = 30 * time.Second
	}

	cwd := opts.Cwd
	if cwd == "" {
		wd, err := os.Getwd()
		if err != nil {
			return nil, fmt.Errorf("toolnexus: acp: resolve cwd: %w", err)
		}
		cwd = wd
	}
	if !filepath.IsAbs(cwd) {
		return nil, fmt.Errorf("toolnexus: acp: Cwd must be absolute, got %q", cwd)
	}

	// The child process is spawned with a background context deliberately —
	// NOT ctx — so it is never killed by the caller cancelling the setup
	// context after LoadACP has returned. ctx only bounds the handshake below.
	cmd := exec.Command(opts.Command, opts.Args...)
	if len(opts.Env) > 0 {
		cmd.Env = append(os.Environ(), opts.Env...)
	}
	stdin, err := cmd.StdinPipe()
	if err != nil {
		return nil, fmt.Errorf("toolnexus: acp: stdin pipe: %w", err)
	}
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return nil, fmt.Errorf("toolnexus: acp: stdout pipe: %w", err)
	}
	if err := cmd.Start(); err != nil {
		return nil, fmt.Errorf("toolnexus: acp: start %q: %w", opts.Command, err)
	}

	c := &ACPClient{
		cmd:               cmd,
		stdin:             stdin,
		stdout:            bufio.NewScanner(stdout),
		pending:           make(map[string]chan acpMsg),
		permissionTimeout: permissionTimeout,
		requestTimeout:    requestTimeout,
	}
	c.stdout.Buffer(make([]byte, 0, 64*1024), 16*1024*1024)
	go c.readLoop()

	if _, err := c.call(ctx, "initialize", map[string]any{
		"protocolVersion": protocolVersion,
		"clientCapabilities": map[string]any{
			"fs": map[string]any{"readTextFile": false, "writeTextFile": false},
		},
	}); err != nil {
		_ = c.Close()
		return nil, fmt.Errorf("toolnexus: acp: initialize: %w", err)
	}

	res, err := c.call(ctx, "session/new", map[string]any{"cwd": cwd, "mcpServers": []any{}})
	if err != nil {
		_ = c.Close()
		return nil, fmt.Errorf("toolnexus: acp: session/new: %w", err)
	}
	var sn struct {
		SessionID string `json:"sessionId"`
	}
	if err := json.Unmarshal(res, &sn); err != nil || sn.SessionID == "" {
		_ = c.Close()
		return nil, fmt.Errorf("toolnexus: acp: session/new: no sessionId in response")
	}
	c.sessionID = sn.SessionID

	if opts.Mode != "" {
		if _, err := c.call(ctx, "session/set_mode", map[string]any{
			"sessionId": c.sessionID,
			"modeId":    opts.Mode,
		}); err != nil {
			_ = c.Close()
			return nil, fmt.Errorf("toolnexus: acp: session/set_mode: %w", err)
		}
	}

	return c, nil
}

// ---------------------------------------------------------------------------
// readLoop: demultiplex everything arriving on the child's stdout.
// ---------------------------------------------------------------------------

func (c *ACPClient) readLoop() {
	for c.stdout.Scan() {
		line := c.stdout.Bytes()
		if len(strings.TrimSpace(string(line))) == 0 {
			continue
		}
		var msg acpMsg
		if err := json.Unmarshal(line, &msg); err != nil {
			continue
		}

		switch {
		case msg.Method == "" && len(msg.ID) > 0:
			// A reply to one of OUR requests (initialize / session/new /
			// session/set_mode / session/prompt), matched by id.
			var idStr string
			_ = json.Unmarshal(msg.ID, &idStr)
			c.mu.Lock()
			ch, ok := c.pending[idStr]
			c.mu.Unlock()
			if ok {
				ch <- msg
			}

		case msg.Method == "session/update":
			c.mu.Lock()
			sink := c.updateSink
			c.mu.Unlock()
			if sink != nil {
				sink(msg.Params)
			}

		case msg.Method == "session/request_permission" && len(msg.ID) > 0:
			// Answered inline, from the read loop itself, so it can never be
			// held up behind whatever sendPrompt is doing — an unanswered
			// permission request hangs the turn forever (ADR 0025).
			c.answerPermissionFirstAllow(msg)

		default:
			// Unhandled server->client notification/request; nothing this
			// client needs (e.g. other session/update variants, fs/* calls
			// we declared unsupported in clientCapabilities).
		}
	}
}

func (c *ACPClient) answerPermissionFirstAllow(req acpMsg) {
	var params struct {
		Options []struct {
			OptionID string `json:"optionId"`
			Kind     string `json:"kind"`
		} `json:"options"`
	}
	_ = json.Unmarshal(req.Params, &params)
	chosen := ""
	for _, o := range params.Options {
		if strings.HasPrefix(o.Kind, "allow") {
			chosen = o.OptionID
			break
		}
	}
	result := map[string]any{"outcome": map[string]any{"outcome": "cancelled"}}
	if chosen != "" {
		result = map[string]any{"outcome": map[string]any{"outcome": "selected", "optionId": chosen}}
	}
	c.reply(req.ID, result)
}

func (c *ACPClient) reply(id json.RawMessage, result any) {
	r, _ := json.Marshal(result)
	c.write(acpMsg{JSONRPC: "2.0", ID: id, Result: r})
}

func (c *ACPClient) write(v any) {
	b, err := json.Marshal(v)
	if err != nil {
		return
	}
	c.mu.Lock()
	defer c.mu.Unlock()
	_, _ = c.stdin.Write(b)
	_, _ = c.stdin.Write([]byte("\n"))
}

// call sends one request and waits for its reply, honoring both ctx and
// c.requestTimeout — used for the one-shot setup calls (initialize,
// session/new, session/set_mode), never for session/prompt (see
// sendPrompt, which has its own permission-aware wait).
func (c *ACPClient) call(ctx context.Context, method string, params any) (json.RawMessage, error) {
	c.mu.Lock()
	c.nextID++
	id := fmt.Sprintf("c-%d", c.nextID)
	ch := make(chan acpMsg, 1)
	c.pending[id] = ch
	c.mu.Unlock()

	idRaw, _ := json.Marshal(id)
	p, _ := json.Marshal(params)
	c.write(acpMsg{JSONRPC: "2.0", ID: idRaw, Method: method, Params: p})

	timer := time.NewTimer(c.requestTimeout)
	defer timer.Stop()

	select {
	case msg := <-ch:
		c.mu.Lock()
		delete(c.pending, id)
		c.mu.Unlock()
		if msg.Error != nil {
			return nil, fmt.Errorf("acp error %d: %s", msg.Error.Code, msg.Error.Message)
		}
		return msg.Result, nil
	case <-timer.C:
		c.mu.Lock()
		delete(c.pending, id)
		c.mu.Unlock()
		return nil, fmt.Errorf("acp: %s timed out after %s", method, c.requestTimeout)
	case <-ctx.Done():
		c.mu.Lock()
		delete(c.pending, id)
		c.mu.Unlock()
		return nil, ctx.Err()
	}
}

// ---------------------------------------------------------------------------
// sendPrompt: one session/prompt turn, accumulating only agent_message_chunk.
// ---------------------------------------------------------------------------

func (c *ACPClient) sendPrompt(text string) (string, error) {
	c.mu.Lock()
	c.nextID++
	id := fmt.Sprintf("c-%d", c.nextID)
	replyCh := make(chan acpMsg, 1)
	c.pending[id] = replyCh

	var accMu sync.Mutex
	var clean strings.Builder
	c.updateSink = func(raw json.RawMessage) {
		var upd struct {
			Update struct {
				SessionUpdate string `json:"sessionUpdate"`
				Content       struct {
					Type string `json:"type"`
					Text string `json:"text"`
				} `json:"content"`
			} `json:"update"`
		}
		if err := json.Unmarshal(raw, &upd); err != nil {
			return
		}
		// Only agent_message_chunk forms the reply: agent_thought_chunk and
		// tool-call narration (tool_call / tool_call_update, which carry no
		// "content" field this struct decodes anyway) are dropped entirely —
		// mixing them corrupts structured output (ADR 0025 gate item #3).
		if upd.Update.SessionUpdate != "agent_message_chunk" {
			return
		}
		accMu.Lock()
		clean.WriteString(upd.Update.Content.Text)
		accMu.Unlock()
	}
	c.mu.Unlock()

	idRaw, _ := json.Marshal(id)
	p, _ := json.Marshal(map[string]any{
		"sessionId": c.sessionID,
		"prompt":    []map[string]any{{"type": "text", "text": text}},
	})
	c.write(acpMsg{JSONRPC: "2.0", ID: idRaw, Method: "session/prompt", Params: p})

	timer := time.NewTimer(c.permissionTimeout)
	defer timer.Stop()

	select {
	case msg := <-replyCh:
		c.mu.Lock()
		delete(c.pending, id)
		c.updateSink = nil
		c.mu.Unlock()
		if msg.Error != nil {
			return "", fmt.Errorf("acp error %d: %s", msg.Error.Code, msg.Error.Message)
		}
		accMu.Lock()
		defer accMu.Unlock()
		return clean.String(), nil
	case <-timer.C:
		c.mu.Lock()
		delete(c.pending, id)
		c.updateSink = nil
		c.mu.Unlock()
		return "", &ACPPermissionTimeoutError{Waited: c.permissionTimeout}
	}
}

// ---------------------------------------------------------------------------
// Generate: the seam into InProcessOptions.Generate / CreateInProcessClient.
// ---------------------------------------------------------------------------

// Generate renders the full assembled request (req.Messages, flattened to
// role+content text) as the prompt for this turn, appends the
// acpSupersedesMarker naming the latest user turn, sends exactly one
// session/prompt on the warm session, and returns the accumulated
// agent_message_chunk text. Turns are serialised: only one session/prompt is
// ever in flight at a time on this client.
func (c *ACPClient) Generate(req InProcessRequest) (InProcessResponse, error) {
	prompt := renderACPPrompt(req)

	c.promptMu.Lock()
	defer c.promptMu.Unlock()

	content, err := c.sendPrompt(prompt)
	if err != nil {
		return InProcessResponse{}, err
	}
	return InProcessResponse{Content: content}, nil
}

// renderACPPrompt flattens req.Messages into "role: content" lines (the
// FULL request, per turn — an ACP session is stateful, so sending only the
// delta would make the client a second, shadow copy of conversation state)
// and appends the supersedes marker naming the latest user turn, so a
// stateful agent answers the current request rather than an earlier
// near-duplicate already sitting in its own session history (ADR 0025).
func renderACPPrompt(req InProcessRequest) string {
	var b strings.Builder
	lastUser := ""
	for _, m := range req.Messages {
		role, content := acpFlattenMessage(m)
		if role == "" && content == "" {
			continue
		}
		fmt.Fprintf(&b, "%s: %s\n", role, content)
		if role == "user" && content != "" {
			lastUser = content
		}
	}
	if lastUser == "" && len(req.Messages) > 0 {
		_, lastUser = acpFlattenMessage(req.Messages[len(req.Messages)-1])
	}
	fmt.Fprintf(&b, "\n%s %s", acpSupersedesMarker, lastUser)
	return b.String()
}

// acpFlattenMessage renders one message's role + content as plain text.
// Content is usually a string; anything else (multimodal parts, etc.) is
// JSON-encoded so it still renders as recognizable text rather than being
// silently dropped.
func acpFlattenMessage(m any) (role, content string) {
	mm, ok := m.(map[string]any)
	if !ok {
		return "", ""
	}
	role, _ = mm["role"].(string)
	switch v := mm["content"].(type) {
	case string:
		content = v
	case nil:
		content = ""
	default:
		if b, err := json.Marshal(v); err == nil {
			content = string(b)
		}
	}
	return role, content
}

// ---------------------------------------------------------------------------
// Close
// ---------------------------------------------------------------------------

// Close terminates the child process and is idempotent — calling it more
// than once is a no-op past the first call.
func (c *ACPClient) Close() error {
	c.closeOnce.Do(func() {
		_ = c.stdin.Close()
		if c.cmd.Process != nil {
			_ = c.cmd.Process.Kill()
		}
		c.closeErr = c.cmd.Wait()
	})
	return c.closeErr
}
